### ISSUE TITLE
Broken content when switching between video ZIM files in Chromium extension

### ISSUE THREAD

**Issue by mossroy** [kiwix/kiwix-js#530]
The behavior is the same with kiwix-js 2.5.0 and the latest code.
It happens only in SW mode (but these files are not really working in jQuery mode)

Steps to reproduce : 
- open the kiwix-js extension in Chromium (tested with version 75)
- open the main page of ted_en_global_issues_2018-07.zim : it is displayed properly
- open the main page of dirtybiology_fr_all_2018-06.zim : you can see thumbnails of TED (and they don't work if you click on it because the link does not exist in the ZIM file)
- close and reopen Chromium, and the kiwix-js extension
- open the main page of dirtybiology_fr_all_2018-06.zim : it is displayed properly

I suspect Chromium keeps some content in cache. I suppose these ZIM files have been generated by the same code, so they might try to call the same URLs.

I did not notice this issue in older versions of Chromium.
I think we should try to fix this before releasing a version 2.6 (#525)

**Comment by Jaifroid** [comment]
So we have to send a cache clear command to the Service Worker when we change ZIMs I guess, analogous to the cache clearing we do in jQuery mode (and probably at the same spot in the code).

**Comment by mossroy** [comment]
We did not implement any custom cache in SW mode.
Using the devtools of Chromium, I see that, in the above scenario, Chromium is storing some javascript files (read from the first ZIM file) in its "memory cache" and re-uses them for the second ZIM file. In particular URLs like chrome-extension://donaljnlmapmngakoipdmehbfcioahhk/www/-/JS/data.js , that stores the list of videos displayed in the main page : it's the same URL for both ZIM files.
We're talking about the standard browser cache.

It causes this regression, that needs to be fixed BUT it should be really good news for performance of the SW mode : we now have (for free) some cache on all assets (CSS, JS, images etc), which is what I was hoping from the very beginning. You can easily check that by using the browser extension, browse a few article and see the "memory cache" mention in the network tab of devtools.

Back to this issue, I think of a workaround : we might change the URL inside the iframe, so that the name of the ZIM file (or some unique identifier) would be in it. Like chrome-extension://donaljnlmapmngakoipdmehbfcioahhk/www/ted_en_global_issues_2018-07.zim/-/JS/data.js
This way, the browser can keep some separate cache for each ZIM file (and clear them with its own strategy)

**Comment by Jaifroid** [comment]
Interesting. But is that better than just issuing `caches.delete(/*name*/);` inside Service Worker when the user changes the ZIM? It just needs us to send a command to the SW in the message channel, and handle it in SW. Might be safer and have less memory footprint. We already have the entry point in our code (because we clear the custom cache in jQuery mode).

**Comment by mossroy** [comment]
What is the `caches` variable you are mentioning?

**Comment by Jaifroid** [comment]
> What is the caches variable you are mentioning?

It is `cssCache`. We have a function `resetCssCache()`, which is called when the ZIM file changes. So we could issue a cache reset command to the Service Worker in the same function.

**Comment by mossroy** [comment]
This `cssCache` is not used in SW mode AFAIK.
And it's not the cache involved here : we're talking about a cache implemented and handled by Chromium itself, not a custom implemented in javascript by us. Clearing the `cssCache` would have no effect.

**Comment by Jaifroid** [comment]
Yes of course. I wasn't suggesting we clear this cache for SW mode, but that in this clear cache function, we send a message to the Service Worker to clear its cache. The message would trigger a command similar to `caches.delete(/*name*/);` inside the SW. In practice, according to https://stackoverflow.com/a/55088582/9727685 we might need to do it async.

**Comment by mossroy** [comment]
I understand, but it's not a cache of the ServiceWorker that is involved here (plus we did not implement any).
If you put a breakpoint in function fetchEventListener of service-worker.js, you'll see that it's not called at all for javascript, CSS and image files : only for HTML files.
In the devtools of Chromium/Chrome, in the network tab, if you check "Disable cache", the problem disappears because all files are fetched through the SW.

So far, I did not find how this cache is called and why/when it has been activated by Google. It seems to be a cache in memory, only for some extension files. And this cache seems to be cleared when the extension tab is closed.
It can make sense : extension files are not supposed to be modified except by upgrading the extension. So caching the files in memory can improve performance.
In our case, they are not actual files (the requests are trapped by our SW, and read in the ZIM file), but the memory cache seems to be involved before the SW is triggered.

In any case, this file caching is a great thing for our performance. Adding a prefix should solve content mixing between ZIM files.
The main problem IMHO is the fact that I don't find any documentation on that

**Comment by Jaifroid** [comment]
OK! Thanks for the explanation. I was assuming it was the standard Service Worker cache.
